File: App/Resources/AlgorithmDetails/bufferedstoogesort/bufferedstoogesort.cpp
#include <cstdio>
#include <utility>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};

void printList(int items[], int size) {
  for (int i = 0; i < size; i++) {
    if (i == 0) {
      printf("[%d, ", items[i]);
    } else if (i != size - 1) {
      printf("%d, ", items[i]);
    } else {
      printf("%d]", items[i]);
    }
  }
}

void bufferedStoogeSort(int arr[], int start, int stop) {
  if (stop - start > 1) {
    if (stop - start == 2 && arr[start] > arr[stop - 1]) {
      std::swap(arr[start], arr[stop - 1]);
    }
    if (stop - start > 2) {
      int width = stop - start;
      int third = (width + 2) / 3 + start;
      int twoThird = (2 * width + 2) / 3 + start;
      if (twoThird - third < third) {
        twoThird--;
      }
      if ((width - 2) % 3 == 0) {
        twoThird--;
      }

      bufferedStoogeSort(arr, third, twoThird);
      bufferedStoogeSort(arr, twoThird, stop);

      int left = third;
      int right = twoThird;
      int bufferStart = start;
      while (left < twoThird && right < stop) {
        if (arr[left] > arr[right]) {
          std::swap(arr[bufferStart], arr[right]);
          right++;
        } else {
          std::swap(arr[bufferStart], arr[left]);
          left++;
        }
        bufferStart++;
      }
      while (right < stop) {
        std::swap(arr[bufferStart], arr[right]);
        right++;
        bufferStart++;
      }

      bufferedStoogeSort(arr, twoThird, stop);

      left = twoThird - 1;
      right = stop - 1;
      while (right > left && left >= start) {
        if (arr[left] > arr[right]) {
          for (int i = left; i < right; i++) {
            std::swap(arr[i], arr[i + 1]);
          }
          left--;
        }
        right--;
      }
    }
  }
}

void sort(int arr[], int n) { bufferedStoogeSort(arr, 0, n); }

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}

File: App/Resources/AlgorithmDetails/bufferedstoogesort/bufferedstoogesort.rb
def buffered_stooge_sort(arr, start, stop)
  if stop - start > 1
    if stop - start == 2 && arr[start] > arr[stop - 1]
      arr[start], arr[stop - 1] = arr[stop - 1], arr[start]
    end
    if stop - start > 2
      width = stop - start
      third = (width + 2) / 3 + start
      two_third = (2 * width + 2) / 3 + start
      two_third -= 1 if two_third - third < third
      two_third -= 1 if (width - 2) % 3 == 0

      buffered_stooge_sort(arr, third, two_third)
      buffered_stooge_sort(arr, two_third, stop)

      left = third
      right = two_third
      buffer_start = start
      while left < two_third && right < stop
        if arr[left] > arr[right]
          arr[buffer_start], arr[right] = arr[right], arr[buffer_start]
          right += 1
        else
          arr[buffer_start], arr[left] = arr[left], arr[buffer_start]
          left += 1
        end
        buffer_start += 1
      end
      while right < stop
        arr[buffer_start], arr[right] = arr[right], arr[buffer_start]
        right += 1
        buffer_start += 1
      end

      buffered_stooge_sort(arr, two_third, stop)

      left = two_third - 1
      right = stop - 1
      while right > left && left >= start
        if arr[left] > arr[right]
          (left...right).each do |i|
            arr[i], arr[i + 1] = arr[i + 1], arr[i]
          end
          left -= 1
        end
        right -= 1
      end
    end
  end
end

def sort(arr)
  buffered_stooge_sort(arr, 0, arr.length)
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array
